File: packages-gen/js/lang/common-resource.js
const common_data = require("@xtalk/lang/common-data.js")

const xts = require("@xtalk/lang/common-string.js")

function xt_existsp(){
  return !(null == globalThis["XT"]);
}

function xt_create(){
  return {
    "::":"xt",
    "config":{},
    "spaces":{},
    "hash":{"lookup":new Map(),"counter":0x011c9dc5}
  };
}

function xt_ensure(){
  if(!(null == globalThis["XT"])){
    return globalThis["XT"];
  }
  else{
    globalThis["XT"] = xt_create();
    return globalThis["XT"];
  }
}

function xt_current(){
  if(!(null == globalThis["XT"])){
    return globalThis["XT"];
  }
  else{
    return null;
  }
}

function xt_purge(){
  if(!(null == globalThis["XT"])){
    let out = globalThis["XT"];
    globalThis["XT"] = null;
    return out;
  }
  else{
    return null;
  }
}

function xt_purge_config(){
  let g = xt_ensure();
  let prev = g["config"];
  g["config"] = {};
  return [true,prev];
}

function xt_purge_spaces(){
  let g = xt_ensure();
  let prev = g["spaces"];
  g["spaces"] = {};
  return [true,prev];
}

function xt_lookup_id(obj){
  if(("function" == (typeof obj)) || ((null != obj) && ("object" == (typeof obj)) && !Array.isArray(obj)) || Array.isArray(obj)){
    let {hash} = xt_ensure();
    let {counter,lookup} = hash;
    let hash_id = lookup.get(obj);
    if(null == hash_id){
      hash["counter"] = (1 + counter);
      lookup.set(obj,counter);
      return counter;
    }
    return hash_id;
  }
}

function xt_config_list(){
  let g = xt_ensure();
  let {config} = g;
  return common_data.obj_keys(config);
}

function xt_config_set(module,m){
  let g = xt_ensure();
  let {config} = g;
  let prev = config[module];
  config[module] = m;
  return [true,prev];
}

function xt_config_del(module){
  let g = xt_ensure();
  let {config} = g;
  let prev = config[module];
  delete(config[module]);
  return [true,prev];
}

function xt_config(module){
  let g = xt_ensure();
  let {config} = g;
  return config[module];
}

function xt_space_list(){
  let g = xt_ensure();
  let {spaces} = g;
  return common_data.obj_keys(spaces);
}

function xt_space_del(module){
  let g = xt_ensure();
  let {spaces} = g;
  let prev = spaces[module];
  if(null != prev){
    delete(spaces[module]);
  }
  return [true,prev];
}

function xt_space(module){
  let g = xt_ensure();
  let {spaces} = g;
  let curr = spaces[module];
  if(null == curr){
    curr = {};
    spaces[module] = curr;
  }
  return curr;
}

function xt_space_clear(module){
  let g = xt_ensure();
  let {spaces} = g;
  let prev = spaces[module];
  spaces[module] = {};
  return [true,prev];
}

function xt_item_del(module,key){
  let space = xt_space(module);
  let prev = space[key];
  if(null != prev){
    delete(space[key]);
  }
  return [true,prev];
}

function xt_item_trigger(module,key){
  let space = xt_space(module);
  let prev = space[key];
  if(null != prev){
    let {value,watch} = prev;
    for(let [watch_key,watch_fn] of Object.entries(watch)){
      watch_fn(value,module + "/" + key);
    };
    return common_data.obj_keys(watch);
  }
}

function xt_item_set(module,key,value){
  let space = xt_space(module);
  let prev = space[key];
  if(null == prev){
    prev = {"watch":{}};
    space[key] = prev;
  }
  prev["value"] = value;
  let {watch} = prev;
  for(let [watch_key,watch_fn] of Object.entries(watch)){
    watch_fn(value,module + "/" + key);
  };
  return [true,prev];
}

function xt_item(module,key){
  let space = xt_space(module);
  let curr = space[key];
  if(curr){
    return curr["value"];
  }
  return null;
}

function xt_item_get(module,key,init_fn){
  let space = xt_space(module);
  let curr = space[key];
  if(curr){
    return curr["value"];
  }
  else{
    let value = init_fn();
    space[key] = {"value":value,"watch":{}};
    return value;
  }
}

function xt_var_entry(sym){
  let [module,key] = xts.sym_pair(sym);
  let space = xt_space(module);
  return space[key];
}

function xt_var(sym){
  let [module,key] = xts.sym_pair(sym);
  return xt_item(module,key);
}

function xt_var_set(sym,value){
  let [module,key] = xts.sym_pair(sym);
  if(null == value){
    return xt_item_del(module,key);
  }
  else{
    return xt_item_set(module,key,value);
  }
}

function xt_var_trigger(sym){
  let [module,key] = xts.sym_pair(sym);
  return xt_item_trigger(module,key);
}

function xt_add_watch(sym,watch_key,watch_fn){
  let entry = xt_var_entry(sym);
  if(entry){
    let {watch} = entry;
    watch[watch_key] = watch_fn;
    return true;
  }
  return false;
}

function xt_remove_watch(sym,watch_key){
  let entry = xt_var_entry(sym);
  if(entry){
    let {watch} = entry;
    delete(watch[watch_key]);
    return true;
  }
  return false;
}

module.exports = {
  ["xt_existsp"]:xt_existsp,
  ["xt_create"]:xt_create,
  ["xt_ensure"]:xt_ensure,
  ["xt_current"]:xt_current,
  ["xt_purge"]:xt_purge,
  ["xt_purge_config"]:xt_purge_config,
  ["xt_purge_spaces"]:xt_purge_spaces,
  ["xt_lookup_id"]:xt_lookup_id,
  ["xt_config_list"]:xt_config_list,
  ["xt_config_set"]:xt_config_set,
  ["xt_config_del"]:xt_config_del,
  ["xt_config"]:xt_config,
  ["xt_space_list"]:xt_space_list,
  ["xt_space_del"]:xt_space_del,
  ["xt_space"]:xt_space,
  ["xt_space_clear"]:xt_space_clear,
  ["xt_item_del"]:xt_item_del,
  ["xt_item_trigger"]:xt_item_trigger,
  ["xt_item_set"]:xt_item_set,
  ["xt_item"]:xt_item,
  ["xt_item_get"]:xt_item_get,
  ["xt_var_entry"]:xt_var_entry,
  ["xt_var"]:xt_var,
  ["xt_var_set"]:xt_var_set,
  ["xt_var_trigger"]:xt_var_trigger,
  ["xt_add_watch"]:xt_add_watch,
  ["xt_remove_watch"]:xt_remove_watch
}

File: packages-gen/dart/lang/lib/common-resource.dart
import 'package:xtalk_lang/common-data.dart' as common_data;

import 'package:xtalk_lang/common-string.dart' as xts;

xt_existsp() {
  return !(null == __globals__["XT"]);
}

xt_create() {
  return <dynamic, dynamic>{
    "::":"xt",
    "config":<dynamic, dynamic>{},
    "spaces":<dynamic, dynamic>{},
    "hash":<dynamic, dynamic>{"lookup":<dynamic, dynamic>{},"counter":0x011c9dc5}
  };
}

xt_ensure() {
  if((() {
    var dart_truthy__39902 = !(null == __globals__["XT"]);
    return (null != dart_truthy__39902) && (false != dart_truthy__39902);
  })()){
    return __globals__["XT"];
  }
  else{
    __globals__["XT"] = xt_create();
    return __globals__["XT"];
  }
}

xt_current() {
  if((() {
    var dart_truthy__39901 = !(null == __globals__["XT"]);
    return (null != dart_truthy__39901) && (false != dart_truthy__39901);
  })()){
    return __globals__["XT"];
  }
  else{
    return null;
  }
}

xt_purge() {
  if((() {
    var dart_truthy__39900 = !(null == __globals__["XT"]);
    return (null != dart_truthy__39900) && (false != dart_truthy__39900);
  })()){
    var out = __globals__["XT"];
    __globals__["XT"] = null;
    return out;
  }
  else{
    return null;
  }
}

xt_purge_config() {
  var g = xt_ensure();
  var prev = g["config"];
  g["config"] = <dynamic, dynamic>{};
  return <dynamic>[true,prev];
}

xt_purge_spaces() {
  var g = xt_ensure();
  var prev = g["spaces"];
  g["spaces"] = <dynamic, dynamic>{};
  return <dynamic>[true,prev];
}

xt_lookup_id(obj) {
  if(((obj.runtimeType).toString().contains("Function") || (obj.runtimeType).toString().contains("=>") || (obj).toString().startsWith("Closure")) || (("Map" == (obj.runtimeType).toString()) || (obj.runtimeType).toString().startsWith("_Map") || (obj.runtimeType).toString().startsWith("LinkedMap")) || ((obj.runtimeType).toString().startsWith("List") || (obj.runtimeType).toString().startsWith("_GrowableList"))){
    var hash = xt_ensure()["hash"];
    var counter = hash["counter"];
    var lookup = hash["lookup"];
    var hash_id = lookup[obj];
    if(null == hash_id){
      hash["counter"] = (1 + counter);
      lookup[obj] = counter;
      return counter;
    }
    return hash_id;
  }
}

xt_config_list() {
  var g = xt_ensure();
  var config = g["config"];
  return common_data.obj_keys(config);
}

xt_config_set(module, m) {
  var g = xt_ensure();
  var config = g["config"];
  var prev = config[module];
  config[module] = m;
  return <dynamic>[true,prev];
}

xt_config_del(module) {
  var g = xt_ensure();
  var config = g["config"];
  var prev = config[module];
  config.remove(module);
  return <dynamic>[true,prev];
}

xt_config(module) {
  var g = xt_ensure();
  var config = g["config"];
  return config[module];
}

xt_space_list() {
  var g = xt_ensure();
  var spaces = g["spaces"];
  return common_data.obj_keys(spaces);
}

xt_space_del(module) {
  var g = xt_ensure();
  var spaces = g["spaces"];
  var prev = spaces[module];
  if(null != prev){
    spaces.remove(module);
  }
  return <dynamic>[true,prev];
}

xt_space(module) {
  var g = xt_ensure();
  var spaces = g["spaces"];
  var curr = spaces[module];
  if(null == curr){
    curr = <dynamic, dynamic>{};
    spaces[module] = curr;
  }
  return curr;
}

xt_space_clear(module) {
  var g = xt_ensure();
  var spaces = g["spaces"];
  var prev = spaces[module];
  spaces[module] = <dynamic, dynamic>{};
  return <dynamic>[true,prev];
}

xt_item_del(module, key) {
  var space = xt_space(module);
  var prev = space[key];
  if(null != prev){
    space.remove(key);
  }
  return <dynamic>[true,prev];
}

xt_item_trigger(module, key) {
  var space = xt_space(module);
  var prev = space[key];
  if(null != prev){
    var value = prev["value"];
    var watch = prev["watch"];
    for(var entry_39903 in watch.entries){
      var watch_key = entry_39903.key;
      var watch_fn = entry_39903.value;
      Function.apply((watch_fn as Function),<dynamic>[value,module + "/" + key]);
    };
    return common_data.obj_keys(watch);
  }
}

xt_item_set(module, key, value) {
  var space = xt_space(module);
  var prev = space[key];
  if(null == prev){
    prev = <dynamic, dynamic>{"watch":<dynamic, dynamic>{}};
    space[key] = prev;
  }
  prev["value"] = value;
  var watch = prev["watch"];
  for(var entry_39904 in watch.entries){
    var watch_key = entry_39904.key;
    var watch_fn = entry_39904.value;
    Function.apply((watch_fn as Function),<dynamic>[value,module + "/" + key]);
  };
  return <dynamic>[true,prev];
}

xt_item(module, key) {
  var space = xt_space(module);
  var curr = space[key];
  if((null != curr) && (false != curr)){
    return curr["value"];
  }
  return null;
}

xt_item_get(module, key, init_fn) {
  var space = xt_space(module);
  var curr = space[key];
  if((null != curr) && (false != curr)){
    return curr["value"];
  }
  else{
    var value = Function.apply((init_fn as Function),<dynamic>[]);
    space[key] = <dynamic, dynamic>{"value":value,"watch":<dynamic, dynamic>{}};
    return value;
  }
}

xt_var_entry(sym) {
  var value_39905 = xts.sym_pair(sym);
  var module = value_39905[0];
  var key = value_39905[1];
  var space = xt_space(module);
  return space[key];
}

xt_var(sym) {
  var value_39906 = xts.sym_pair(sym);
  var module = value_39906[0];
  var key = value_39906[1];
  return xt_item(module,key);
}

xt_var_set(sym, value) {
  var value_39907 = xts.sym_pair(sym);
  var module = value_39907[0];
  var key = value_39907[1];
  if(null == value){
    return xt_item_del(module,key);
  }
  else{
    return xt_item_set(module,key,value);
  }
}

xt_var_trigger(sym) {
  var value_39908 = xts.sym_pair(sym);
  var module = value_39908[0];
  var key = value_39908[1];
  return xt_item_trigger(module,key);
}

xt_add_watch(sym, watch_key, watch_fn) {
  var entry = xt_var_entry(sym);
  if((null != entry) && (false != entry)){
    var watch = entry["watch"];
    watch[watch_key] = watch_fn;
    return true;
  }
  return false;
}

xt_remove_watch(sym, watch_key) {
  var entry = xt_var_entry(sym);
  if((null != entry) && (false != entry)){
    var watch = entry["watch"];
    watch.remove(watch_key);
    return true;
  }
  return false;
}